### ISSUE TITLE
Refactor using txaio to support Twisted and asyncio backends

### ISSUE THREAD

**Issue by lthibault** [davebshow/goblin#48]
Goblin's reliance on asyncio means it cannot be used within a swath of popular frameworks built on top of Twisted.  Case in point:  [Scrapy](https://scrapy.org/) spiders can't make use of Goblin.  While this is a problem for me, personally, I think it's also a limiting factor in the widespread adoption of Goblin.

This limitation could be solved by rewriting golbin to use [Autobahn](http://autobahn.ws/python/websocket/examples.html#websocket-examples)'s websocket implementation, which can employ asyncio or Twisted as a backend.  For other asynchronous logic, [txaio](https://github.com/crossbario/txaio) can be used to write coroutines and futures/deferreds in a manner agnostic to the async backend.

Is this something the Goblin team would be interested in doing?

**Comment by davebshow** [comment]
Hi @lthibault. Thanks for your interest in Goblin. The answer to your question is no, we don't want to rewrite Goblin. There are a lot of reasons for this, the main one is that we want to continue using the Python standard library support for asynchronous programming. Eventually, we will probably even remove aiohttp as a dependency.

That said, you are welcome to fork the project, do your refactor (which would be fairly extensive), and maintain a twisted Goblin implementation. I would be happy to answer any questions and do what is necessary to facilitate this (short of writing the code myself).

Another option would be to use the official Gremlin-Python Driver which will be packaged with TinkerPop starting with the 3.2.5. It is a threaded implementation that makes it easy to plug in any websocket client you want by defining a simple wrapper. 

Finally, short of doing either of the above, it should be relatively easy to integrate Goblin with a Twisted app even though they don't run on the same event loop. TBH, I don't have any experience with Twisted, so I can't foresee what potential problems would be, but it seems to me that you could simply run an asyncio event loop in a separate thread, and add Goblin tasks to the thread from your Twisted app. Beyond this, there are a variety of ways you could glue several small applications together to get the kind of integration you are looking for.

Anyway, let me know what you decide.

**Comment by lthibault** [comment]
Hi David,

I somewhat jumped the gun on this issue — it turns out Twisted now has an `asyncioractor` and supports `async` and `await` on deferred function calls.  This means it’s actually possible to integrate asyncio code within a Twisted project.

In short:  this issue is, in fact, a non-issue :)

Thanks for the fantastic project!

**Comment by davebshow** [comment]
Hey that is great news. Kudos to the Twisted team.

A quick question for you: are you more interested in the OGM, the Driver, or the async Gremlin-Python implementation? I'm asking because due to the relatively complex TP ecosystem, I am thinking about extracting the Driver/Gremlin-Python implementation to the `aiogremlin` module namespace, and leaving the OGM functionality with Goblin. At any rate, in that case Goblin will still work the same, and I will try to make sure that imports etc. aren't broken. I'm going to close, but please reply to the question if you have a minute.

**Comment by lthibault** [comment]
I know!  They work fast!

It’s primarily the OGM that we're interested in.  As a matter of fact, the reliance on Gremlin-Python turns out to be a problem, as this library relies on Gremlin 3.2.3, which in turn requires Titan to be compiled from source with an unmerged PR.  It’s not a deal-breaker for us, but it’s a pain point, and we’d actually prefer if the OGM could be used in conjunction with the standard groovy language driver.

As for networking concerns, the main issue is that Goblin has to be usable on top of Twisted, since we intend to use it within Scrapy spiders.  This seems pretty straightforward on my end.

In other words:  in a perfect world, the OGM would be factored in such a way that it is agnostic to the networking layer and the language driver.  Is this essentially what you’re suggesting?  If so, color me very interested!  :)

**Comment by davebshow** [comment]
A couple things:

> As a matter of fact, the reliance on Gremlin-Python turns out to be a problem, as this library relies on Gremlin 3.2.3, which in turn requires Titan to be compiled from source with an unmerged PR.

Well, yes and no. The Gremlin-Python bundled with Goblin includes a GraphSON1MessageSerializer so you can use it will older server versions, the one included in the Titan 1.1 should work fine with Goblin. If you are interested, this [repo](https://github.com/petteraas/docker-titan) helps you build branches of Titan with certain versions of TinkerPop. In general, I would highly recommend Titan 1.1 over Titan 1.0; however, it is getting to the point now that I would say if you are developing an app that you want to use for more than a month or two, I would switch over to JanusGraph, which is Titan's new name now that it is being backed by Linux Foundation, IBM, etc. However, if you still want to use Titan 1.1 and want to build it yourself, you should be able to do something like:

```
git clone https://github.com/thinkaurelius/titan -b titan11 titan11
cd titan11
mvn clean install -DskipTests=true -Dgpg.skip=true -Paurelius-release
unzip titan-dist/titan-dist-hadoop-2/target/titan-1.1.0-SNAPSHOT-hadoop2.zip
```

> It’s not a deal-breaker for us, but it’s a pain point, and we’d actually prefer if the OGM could be used in conjunction with the standard groovy language driver.

I'm not really sure what you mean here...There is a Java Driver packaged with TinkerPop, and there are Java OGMs available...But, in general, you won't be using a Python OGM with a JVM driver.

> In other words:  in a perfect world, the OGM would be factored in such a way that it is agnostic to the networking layer and the language driver.

Well, right now I am extracting a PEP 492 compatible version of Gremlin-Python from Goblin that will be packaged under the aiogremlin namespace. This includes a async driver, which supports plugable websocket client implementations.

**Comment by lthibault** [comment]
Hi David,

Thanks for the information and for the suggestions.  Why exactly do you suggest moving over to JanusGraph?  I take it JanusGraph is fully compatible with Titan (and therefore Gremlin) ?

>I'm not really sure what you mean here...There is a Java Driver packaged with TinkerPop, and there are Java OGMs available...But, in general, you won't be using a Python OGM with a JVM driver.

I think I may be confused about Goblin’s architecture.  What exactly do you mean by “Driver" in your previous email?

Thanks
